ssl/s23_clnt.c

```diff
@@ -498,8 +480,21 @@ static int ssl23_client_hello(SSL *s)
             memcpy(p, s->s3->client_random, SSL3_RANDOM_SIZE);
             p += SSL3_RANDOM_SIZE;
 
-            /* Session ID (zero since there is no reuse) */
-            *(p++) = 0;
+            /* Session ID */
+            if (s->new_session || s->session == NULL)
+                i = 0;
+            else
+                i = s->session->session_id_length;
```

Comment:
I see this in various places, but I have to wonder that new_session is correct or not, both here and in the master branch.  At least the ssl_get_new_session() call doesn't seem to set that.